subsys/mgmt/mcumgr/grp/fs_mgmt/src/fs_mgmt.c

```diff
@@ -354,44 +384,139 @@ static int fs_mgmt_file_upload(struct smp_streamer *ctxt)
 		if (len == ULLONG_MAX) {
 			return MGMT_ERR_EINVAL;
 		}
+	}
+
+	if (k_sem_take(&fs_mgmt_ctxt.lock_sem, FILE_SEMAPHORE_MAX_TAKE_TIME)) {
+		return MGMT_ERR_EBUSY;
+	}
+
+	/* Check if this upload is already in progress */
+	if (ctxt->smpt != fs_mgmt_ctxt.transport ||
+	    fs_mgmt_ctxt.state != STATE_UPLOAD ||
+	    strcmp(file_name, fs_mgmt_ctxt.path)) {
+		fs_mgmt_cleanup();
+	}
 
-		fs_mgmt_ctxt.uploading = true;
+	/* Open new file */
+	if (fs_mgmt_ctxt.state == STATE_NO_UPLOAD_OR_DOWNLOAD) {
 		fs_mgmt_ctxt.off = 0;
+		fs_file_t_init(&fs_mgmt_ctxt.file);
+		rc = fs_open(&fs_mgmt_ctxt.file, file_name, FS_O_CREATE | FS_O_WRITE);
+
+		if (rc != 0) {
+			rc = MGMT_ERR_ENOENT;
+			goto end;
+		}
+
+		strcpy(fs_mgmt_ctxt.path, file_name);
+		fs_mgmt_ctxt.state = STATE_UPLOAD;
+		fs_mgmt_ctxt.transport = ctxt->smpt;
+	}
+
+	if (off == 0) {
+		/* Store the uploaded file size from the first packet, this will allow
+		 * closing the file when the full upload has finished, however the file
+		 * will remain opened if the upload state is lost. It will, however,
+		 * still be closed automatically after a timeout.
+		 */
 		fs_mgmt_ctxt.len = len;
-	} else {
-		if (!fs_mgmt_ctxt.uploading) {
-			return MGMT_ERR_EINVAL;
+		rc = fs_mgmt_filelen(file_name, &existing_file_size);
+
+		if (rc != 0) {
+			rc = MGMT_ERR_EUNKNOWN;
+			fs_mgmt_cleanup();
+			goto end;
 		}
+	} else if (fs_mgmt_ctxt.off == 0) {
+		rc = fs_mgmt_filelen(file_name, &fs_mgmt_ctxt.off);
 
-		if (off != fs_mgmt_ctxt.off) {
-			/* Invalid offset.  Drop the data and send the expected offset. */
-			return fs_mgmt_file_rsp(zse, MGMT_ERR_EINVAL, fs_mgmt_ctxt.off);
+		if (rc != 0) {
+			rc = MGMT_ERR_EUNKNOWN;
+			fs_mgmt_cleanup();
+			goto end;
 		}
 	}
 
-	new_off = fs_mgmt_ctxt.off + file_data.len;
-	if (new_off > fs_mgmt_ctxt.len) {
-		/* Data exceeds image length. */
-		return MGMT_ERR_EINVAL;
+	/* Verify that the data offset matches the expected offset (i.e. current size of file) */
+	if (off != fs_mgmt_ctxt.off) {
```

Comment:
IMHO client should be notified that the file it is writing is altered somehow so client is appending to something that is different then it started originally to upload.

Reply:
I did not mean to add different parm. What I meant is that in case when real file len differs you ask for the `off` equal to the real file len. If client does not hold track of how much it uploaded, then it just tries to upload the offset it has been requested to, but if it does, then the client may now error out, because something is not ok.
Now the client can, for example, request sha of the file up to the offset that has been requested and figure whether this is still the file it has been uploading.

Reply:
This places an unknown bias code flow onto clients which breaks compatibility with the existing upload method. I would compromise on returning an error instead, since if the file has changed then the user will need to reupload the file again anyway, will rework to do that.

Reply:
Have added a new commit, see what you think of that - it's a halfway step. I think really the system would need to be enhanced with a returning an error, if rc is used strictly for the SMP error code then it should always return ok (or a new error code e.g. ECHECKCMDRC, in places where it returns eunknown now, and a new parameter added for the groups themselves, e.g. "cmdrc" which has the status of the command, so for fs_mgmt this would be things like file_open_failed, file_seek_failed, file_does_not_exist, file_has_changed, etc. which would completely break the go application. I don't have any issue in breaking compatibility with the go application but would not want to do that for this release, and we would need at least some other implementation (e.g. python) that people can use instead of that janky go code to test the new functionality - thoughts?

Reply:
> This places an unknown bias code flow onto clients which breaks compatibility with the existing upload method. I would compromise on returning an error instead, since if the file has changed then the user will need to reupload the file again anyway, will rework to do that.

Yes. That what I meant. The original behaviour would return an error if unexpected offset has been received by the smp ervice, whether it has been lower than expected or higher than expected.
In case when we request lower, user does not know whether file has changed or frame has been dropped.
When we request higher than sending app should not get confused, unless it has bad error handling.
In both cases, when app breaks it is not our fault, we can request any offset we can.
We do not have to report, currently, anything more than we originally did.
We are providing command that allows user to verify sha of uploaded file and if user really cares about contents of the file, the command should be used.

Reply:
@nordicjm to check code as per above comment & update code

Reply:
Should be fixed, please recheck @de-nordic